cpp/src/join/join.cu

```diff
@@ -314,11 +314,15 @@ std::unique_ptr<experimental::table> construct_join_output_df(
       auto common_from_right = experimental::detail::gather(right.select(right_common_col),
                                                             complement_indices.second.begin(),
                                                             complement_indices.second.end(),
-                                                            nullify_out_of_bounds);
+                                                            nullify_out_of_bounds,
+                                                            mr,
```

Comment:
`common_from_right` and `common_from_left` here should be using the default resource instead of `mr` since they're only temporary, not returned columns. It _is_ nice that they have the `stream` parameter hooked up though...

Reply:
I see. I was naively assuming that the subsequent `concatenate` call merely groups `common_from_left` and `common_from_right` without copying them. I just pushed another change that treats those columns as temporary and passes the user-provided resource to `concatenate` instead (which is necessary because `common_table` is returned to the user). @trevorsm7 could you make another pass on it and resolve this conversation if you're satisfied?

Reply:
Ah, good catch passing `mr` to `concatenate`. I think you should pass `stream` there as well though.

Reply:
I don't think `concatenate` takes a stream argument: https://github.com/rapidsai/cudf/blob/branch-0.14/cpp/include/cudf/concatenate.hpp#L86-L88

Reply:
fyi, `concatenate` isn't the only function that doesn't take a stream argument, and I believe there is an ongoing discussion about making synchronization behavior consistent: https://github.com/rapidsai/cudf/issues/4511

Reply:
You can use the detail concatenate https://github.com/rapidsai/cudf/blob/5f95c9745e068d51c2a4a59d6fd7ce9f94a91b9a/cpp/include/cudf/detail/concatenate.cuh#L73-L76
The non-detail is really just a thin wrapper around this that hides the stream parameter (for the reasons mentioned in #4511) https://github.com/rapidsai/cudf/blob/5f95c9745e068d51c2a4a59d6fd7ce9f94a91b9a/cpp/src/copying/concatenate.cu#L365-L370

Reply:
Sorry that this got a little messy. There are probably quite a few places where the stream parameters could be hooked up to detail APIs, and it doesn't make a difference since we currently only use the default stream anyway, but we should probably be fixing these as we come across them otherwise they'll end up being bugs down the road when we do support strings.

Reply:
Are you sure this is the function being invoked here? I thought that `gather` returns a table , and therefore the following `concatenate` call would pick this overload: https://github.com/rapidsai/cudf/blob/5f95c9745e068d51c2a4a59d6fd7ce9f94a91b9a/cpp/include/cudf/concatenate.hpp#L86-L88

And I wouldn't call this function a "thin" wrapper: https://github.com/rapidsai/cudf/blob/5f95c9745e068d51c2a4a59d6fd7ce9f94a91b9a/cpp/src/copying/concatenate.cu#L373-L394

Do you still think that I can pass a stream here?

Reply:
> Sorry that this got a little messy. There are probably quite a few places where the stream parameters could be hooked up to detail APIs but aren't, and it doesn't make a difference since we currently only use the default stream anyway, but we should probably be fixing these as we come across them otherwise they'll end up being bugs down the road when we do support streams.
No worries. I've encountered such places quite a few times, and I don't mind fixing them myself. I just think that the `concatenate` overload that I think is being used here better be fixed to take a stream argument, so that we don't need to inline the whole body.

Reply:
Note that the one you linked is an overload for tables instead of columns
` std::unique_ptr<table> concatenate(std::vector<table_view>`
The correct one is here https://github.com/rapidsai/cudf/blob/5f95c9745e068d51c2a4a59d6fd7ce9f94a91b9a/cpp/src/copying/concatenate.cu#L365-L370

Reply:
Wait, maybe you're right... are the inputs tables here?

Reply:
Oh, I see. I thought `common_from_right` and `common_from_left` were columns. Didn't notice since they were `auto`. Hm.

Reply:
Ok, I think we should open another issue to add a detail/stream variant for the table overload of concatenate and reference this PR for this to be fixed later. I'll remove my change request so you can merge this PR as-is.